File: backend/export.go
package main

import (
	"context"
	"net/http"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strings"
)

type exportRequest struct {
	URL         string   `json:"url"`
	Ignore      []string `json:"ignore"`
	Include     []string `json:"include"`
	Select      []string `json:"select"`
	All         bool     `json:"all"`
	Destination string   `json:"destination"`
	Format      string   `json:"format"`
	PathMode    string   `json:"path_mode"`
}

type exportResult struct {
	Format    string `json:"format"`
	Path      string `json:"path"`
	FileCount int    `json:"file_count"`
	TotalSize int64  `json:"total_size"`
}

func exportToHost(ctx context.Context, files []FileEntry, client *GitHubClient, config Config, request exportRequest, repoName string) (exportResult, error) {
	if config.DownloadRoot == "" {
		return exportResult{}, hostExportUnavailableError()
	}
	if err := os.MkdirAll(config.DownloadRoot, 0o755); err != nil {
		return exportResult{}, &APIError{Status: http.StatusServiceUnavailable, Message: "宿主机导出目录不可用", Hint: "请检查 DOWNLOAD_ROOT 对应的 Docker 挂载目录权限。", Code: "host_export_unavailable", Cause: err}
	}
	if !isWritableDirectory(config.DownloadRoot) {
		return exportResult{}, hostExportUnavailableError()
	}
	root, err := filepath.Abs(config.DownloadRoot)
	if err != nil {
		return exportResult{}, &APIError{Status: 500, Message: "could not resolve DOWNLOAD_ROOT", Cause: err}
	}
	root, err = filepath.EvalSymlinks(root)
	if err != nil {
		return exportResult{}, &APIError{Status: 500, Message: "could not resolve DOWNLOAD_ROOT", Cause: err}
	}
	destination := strings.TrimSpace(request.Destination)
	if destination == "" {
		destination = repoName
	}
	relative, err := safeRelativePath(destination)
	if err != nil {
		return exportResult{}, err
	}
	format := request.Format
	if format == "" {
		format = "zip"
	}
	switch format {
	case "folder":
		pathMode := request.PathMode
		if pathMode == "" {
			pathMode = "original"
		}
		if pathMode != "original" && pathMode != "smart" {
			return exportResult{}, invalidError("unsupported folder path mode: " + pathMode)
		}
		return exportFolder(ctx, root, relative, files, client, config, normalizeSelection(request.Select, request.All), pathMode)
	case "zip":
		return exportZIP(ctx, root, relative, files, client, config)
	default:
		return exportResult{}, invalidError("unsupported export format: " + format)
	}
}

func normalizeSelection(selection []string, all bool) []string {
	if all {
		return nil
	}
	return selection
}

func hostExportUnavailableError() *APIError {
	return &APIError{
		Status:  http.StatusServiceUnavailable,
		Message: "宿主机导出目录不可写",
		Hint:    "请检查 DOWNLOAD_ROOT 对应的 Docker 挂载目录权限。",
		Code:    "host_export_unavailable",
	}
}

func isWritableDirectory(directory string) bool {
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return false
	}
	testFile, err := os.CreateTemp(directory, ".zzz-write-test-*")
	if err != nil {
		return false
	}
	testPath := testFile.Name()
	if err := testFile.Close(); err != nil {
		_ = os.Remove(testPath)
		return false
	}
	return os.Remove(testPath) == nil
}

func exportFolder(ctx context.Context, root, destination string, files []FileEntry, client *GitHubClient, config Config, selection []string, pathMode string) (exportResult, error) {
	output, err := safeJoin(root, destination)
	if err != nil {
		return exportResult{}, err
	}
	if err := ensureNoSymlink(root, output); err != nil {
		return exportResult{}, err
	}
	if err := os.MkdirAll(output, 0o755); err != nil {
		return exportResult{}, &APIError{Status: 500, Message: "could not create output folder", Cause: err}
	}
	stripPrefix := ""
	if pathMode == "smart" {
		stripPrefix = smartFolderPrefix(files, selection)
	}
	var totalSize int64
	for _, file := range files {
		content, err := client.downloadFile(ctx, file)
		if err != nil {
			return exportResult{}, err
		}
		totalSize += int64(len(content))
		if totalSize > config.MaxTotalSize {
			return exportResult{}, &APIError{Status: 413, Message: "total downloaded content is too large"}
		}
		relativePath := file.Path
		if stripPrefix != "" {
			relativePath = strings.TrimPrefix(relativePath, stripPrefix+"/")
		}
		outputPath, err := safeJoin(output, relativePath)
		if err != nil {
			return exportResult{}, err
		}
		if err := ensureNoSymlink(root, outputPath); err != nil {
			return exportResult{}, err
		}
		if err := os.MkdirAll(filepath.Dir(outputPath), 0o755); err != nil {
			return exportResult{}, &APIError{Status: 500, Message: "could not create output folder", Cause: err}
		}
		if err := os.WriteFile(outputPath, content, 0o644); err != nil {
			return exportResult{}, &APIError{Status: 500, Message: "could not write output file", Cause: err}
		}
	}
	return exportResult{Format: "folder", Path: filepath.ToSlash(destination), FileCount: len(files), TotalSize: totalSize}, nil
}

func smartFolderPrefix(files []FileEntry, selection []string) string {
	if len(files) == 0 || len(selection) == 0 {
		return ""
	}

	parents := make([]string, 0, len(selection))
	for _, root := range reduceSelection(selection) {
		parent := path.Dir(root)
		if parent == "." {
			parent = ""
		}
		parents = append(parents, parent)
	}
	return commonPathPrefix(parents)
}

func reduceSelection(selection []string) []string {
	unique := make(map[string]struct{}, len(selection))
	for _, selected := range selection {
		selected = strings.Trim(selected, "/")
		if selected != "" {
			unique[selected] = struct{}{}
		}
	}
	paths := make([]string, 0, len(unique))
	for selected := range unique {
		paths = append(paths, selected)
	}
	sort.Slice(paths, func(i, j int) bool {
		if len(paths[i]) == len(paths[j]) {
			return paths[i] < paths[j]
		}
		return len(paths[i]) < len(paths[j])
	})
	reduced := make([]string, 0, len(paths))
	for _, candidate := range paths {
		covered := false
		for _, parent := range reduced {
			if candidate == parent || strings.HasPrefix(candidate, parent+"/") {
				covered = true
				break
			}
		}
		if !covered {
			reduced = append(reduced, candidate)
		}
	}
	return reduced
}

func commonPathPrefix(paths []string) string {
	if len(paths) == 0 || paths[0] == "" {
		return ""
	}
	parts := strings.Split(paths[0], "/")
	for _, value := range paths[1:] {
		if value == "" {
			return ""
		}
		other := strings.Split(value, "/")
		limit := len(parts)
		if len(other) < limit {
			limit = len(other)
		}
		index := 0
		for index < limit && parts[index] == other[index] {
			index++
		}
		parts = parts[:index]
		if len(parts) == 0 {
			return ""
		}
	}
	return strings.Join(parts, "/")
}

func exportZIP(ctx context.Context, root, destination string, files []FileEntry, client *GitHubClient, config Config) (exportResult, error) {
	if filepath.Ext(destination) == "" {
		destination += ".zip"
	}
	output, err := safeJoin(root, destination)
	if err != nil {
		return exportResult{}, err
	}
	if err := ensureNoSymlink(root, output); err != nil {
		return exportResult{}, err
	}
	if err := os.MkdirAll(filepath.Dir(output), 0o755); err != nil {
		return exportResult{}, &APIError{Status: 500, Message: "could not create output folder", Cause: err}
	}
	zipData, err := createZIP(ctx, files, client, config, filepath.Dir(output))
	if err != nil {
		return exportResult{}, err
	}
	if err := os.Chmod(zipData.path, 0o644); err != nil {
		_ = os.Remove(zipData.path)
		return exportResult{}, &APIError{Status: 500, Message: "could not set ZIP file permissions", Cause: err}
	}
	if err := os.Rename(zipData.path, output); err != nil {
		_ = os.Remove(zipData.path)
		return exportResult{}, &APIError{Status: 500, Message: "could not write ZIP file", Cause: err}
	}
	return exportResult{Format: "zip", Path: filepath.ToSlash(destination), FileCount: len(files), TotalSize: zipData.size}, nil
}

func safeRelativePath(value string) (string, error) {
	value = strings.TrimSpace(strings.ReplaceAll(value, "\\", "/"))
	if value == "" || strings.HasPrefix(value, "/") {
		return "", invalidError("output path must be relative and cannot contain '..'")
	}
	for _, part := range strings.Split(value, "/") {
		if part == ".." {
			return "", invalidError("output path must be relative and cannot contain '..'")
		}
	}
	clean := pathCleanSlash(value)
	if clean == "." || clean == ".." || strings.HasPrefix(clean, "../") {
		return "", invalidError("output path must be relative and cannot contain '..'")
	}
	return filepath.FromSlash(clean), nil
}

func pathCleanSlash(value string) string {
	parts := make([]string, 0)
	for _, part := range strings.Split(value, "/") {
		switch part {
		case "", ".":
		case "..":
			if len(parts) > 0 {
				parts = parts[:len(parts)-1]
			} else {
				return ".."
			}
		default:
			parts = append(parts, part)
		}
	}
	return strings.Join(parts, "/")
}

func safeJoin(root, relative string) (string, error) {
	joined := filepath.Join(root, relative)
	rel, err := filepath.Rel(root, joined)
	if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return "", invalidError("output path escapes DOWNLOAD_ROOT")
	}
	return joined, nil
}

func ensureNoSymlink(root, target string) error {
	rel, err := filepath.Rel(root, target)
	if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return invalidError("output path escapes DOWNLOAD_ROOT")
	}
	current := root
	for _, part := range strings.Split(rel, string(filepath.Separator)) {
		if part == "." || part == "" {
			continue
		}
		current = filepath.Join(current, part)
		info, err := os.Lstat(current)
		if err == nil && info.Mode()&os.ModeSymlink != 0 {
			return invalidError("output path cannot be a symbolic link")
		}
		if err != nil && !os.IsNotExist(err) {
			return &APIError{Status: 500, Message: "could not inspect output path", Cause: err}
		}
	}
	return nil
}

File: backend/src/export.rs
use std::path::{Component, Path, PathBuf};

use serde::Serialize;
use tokio::fs;

use crate::{
    config::AppConfig,
    error::AppError,
    github::client::{FileEntry, GitHubClient},
    zip::create_zip,
};

#[derive(Debug, Clone, Copy)]
pub enum ExportFormat {
    Folder,
    Zip,
}

#[derive(Debug, Serialize)]
pub struct ExportResult {
    pub format: &'static str,
    pub path: String,
    pub file_count: usize,
    pub total_size: u64,
}

pub async fn export_to_host(
    files: &[FileEntry],
    client: &GitHubClient,
    config: &AppConfig,
    destination: Option<&str>,
    format: ExportFormat,
    repo_name: &str,
) -> Result<ExportResult, AppError> {
    let root = config
        .download_root
        .as_ref()
        .ok_or_else(|| AppError::Configuration("DOWNLOAD_ROOT is not configured".into()))?;
    fs::create_dir_all(root)
        .await
        .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
    let root = fs::canonicalize(root)
        .await
        .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
    let destination = destination
        .filter(|value| !value.trim().is_empty())
        .unwrap_or(repo_name);
    let relative_destination = safe_relative_path(destination)?;

    match format {
        ExportFormat::Folder => {
            let output = root.join(&relative_destination);
            fs::create_dir_all(&output)
                .await
                .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
            let output = ensure_inside_root(&root, &output).await?;
            let mut total_size = 0_u64;
            for file in files {
                let content = client.download_file(&file.content_url, file.size).await?;
                total_size = total_size.saturating_add(content.len() as u64);
                let path = safe_join(&output, &file.path)?;
                if let Some(parent) = path.parent() {
                    fs::create_dir_all(parent)
                        .await
                        .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
                    ensure_inside_root(&output, parent).await?;
                }
                reject_symlink(&path).await?;
                fs::write(&path, content)
                    .await
                    .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
            }
            Ok(ExportResult {
                format: "folder",
                path: relative_destination.to_string_lossy().into_owned(),
                file_count: files.len(),
                total_size,
            })
        }
        ExportFormat::Zip => {
            let mut relative_file = relative_destination;
            if relative_file.extension().is_none() {
                relative_file.set_extension("zip");
            }
            let output = root.join(&relative_file);
            if let Some(parent) = output.parent() {
                fs::create_dir_all(parent)
                    .await
                    .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
                ensure_inside_root(&root, parent).await?;
            }
            let zip = create_zip(files, client, config).await?;
            reject_symlink(&output).await?;
            fs::write(&output, &zip)
                .await
                .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
            Ok(ExportResult {
                format: "zip",
                path: relative_file.to_string_lossy().into_owned(),
                file_count: files.len(),
                total_size: zip.len() as u64,
            })
        }
    }
}

fn safe_relative_path(value: &str) -> Result<PathBuf, AppError> {
    let path = Path::new(value.trim());
    if path.as_os_str().is_empty()
        || path.components().any(|component| {
            matches!(
                component,
                Component::RootDir | Component::Prefix(_) | Component::ParentDir
            )
        })
    {
        return Err(AppError::InvalidUrl(
            "output path must be relative and cannot contain '..'".into(),
        ));
    }
    Ok(path.to_path_buf())
}

fn safe_join(root: &Path, relative: &str) -> Result<PathBuf, AppError> {
    let relative = safe_relative_path(relative)?;
    Ok(root.join(relative))
}

async fn ensure_inside_root(root: &Path, path: &Path) -> Result<PathBuf, AppError> {
    let canonical = fs::canonicalize(path)
        .await
        .map_err(|error| AppError::Other(anyhow::Error::new(error)))?;
    if !canonical.starts_with(root) {
        return Err(AppError::InvalidUrl(
            "output path escapes DOWNLOAD_ROOT".into(),
        ));
    }
    Ok(canonical)
}

async fn reject_symlink(path: &Path) -> Result<(), AppError> {
    if let Ok(metadata) = fs::symlink_metadata(path).await
        && metadata.file_type().is_symlink()
    {
        return Err(AppError::InvalidUrl(
            "output path cannot be a symbolic link".into(),
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::safe_relative_path;

    #[test]
    fn accepts_relative_output_paths() {
        assert_eq!(
            safe_relative_path("ebooks/2025").unwrap().to_string_lossy(),
            "ebooks/2025"
        );
    }

    #[test]
    fn rejects_paths_that_can_escape_the_root() {
        assert!(safe_relative_path("../outside").is_err());
        assert!(safe_relative_path("/outside").is_err());
    }
}
